### PR TITLE
Files

### PR DESCRIPTION
**Pull request by eatkins**
This pr was a result of my working to make the [Ammonite-Ops](http://ammonite.io/#Ammonite-Ops) library work with scala native. I made a number of changes that improved performance or updated the semantics of scala native to more closely match the jvm.

### CODE REVIEW

**Comment by densh** [comment]
@eatkins There has been some conflicting changes due to #1129 being merged first. Please rebase your changes on top of current master.

File: javalib/src/main/scala/java/nio/file/Files.scala [hunk 1]
```diff
@@ -315,19 +347,25 @@ object Files {
...
+    new WrappedScalaStream(
+      FileHelpers.list(dir.toString, (n, _) => dir.resolve(n)).toStream,
+      None)
 
   def move(source: Path, target: Path, options: Array[CopyOption]): Path = {
-    copy(source, target, options)
-    delete(source)
+    if (!exists(source.toAbsolutePath, Array.empty))
+      throw new NoSuchFileException(source.toString)
+    if (!exists(target.toAbsolutePath, Array.empty) || options.contains(
```
Comment (densh) [inline javalib/src/main/scala/java/nio/file/Files.scala]: Curlies around the first then clause, and then `else if` afterwards would read better.

File: javalib/src/main/scala/java/nio/file/Files.scala [hunk 2]
```diff
@@ -391,17 +429,22 @@ object Files {
...
+      var read   = 0
+      while ({
+        read = unistd.read(fd, bytes.at(offset), len - offset);
+        read != -1 && (offset + read) < len
+      }) {
+        offset += read
+      }
+      if (read == -1) throw UnixException(path.toString, errno.errno)
+      bytes.asInstanceOf[Array[Byte]]
+    } finally fcntl.close(fd)
```
Comment (densh) [inline javalib/src/main/scala/java/nio/file/Files.scala]: Please add curlies around finally clause.

File: javalib/src/main/scala/scala/scalanative/nio/fs/FileHelpers.scala [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import scalanative.posix.dirent._
+import scalanative.posix.{errno => e, fcntl, unistd}, e._, unistd.access
+import scalanative.native._, stdlib._, stdio._, string._
+import scala.collection.mutable.UnrolledBuffer
+import scala.reflect.ClassTag
+import java.io.{File, IOException}
+import fcntl._
+
+object FileHelpers {
+  case class Dirent(name: String, tpe: CShort)
```
Comment (densh) [inline javalib/src/main/scala/scala/scalanative/nio/fs/FileHelpers.scala]: `final case class`.

File: javalib/src/main/scala/scala/scalanative/nio/fs/FileHelpers.scala [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        result = genTempFile(prefix, newSuffix, tmpDir)
+      } while (!createNewFile(result.toString))
+      result
+    }
+
+  def exists(path: String): Boolean =
+    Zone { implicit z =>
+      access(toCString(path), fcntl.F_OK) == 0
+    }
+  private[this] lazy val random = new scala.util.Random()
```
Comment (densh) [inline javalib/src/main/scala/scala/scalanative/nio/fs/FileHelpers.scala]: We tend to put all the state early in the object/class.

File: javalib/src/main/scala/scala/scalanative/nio/fs/NativePosixFileAttributeView.scala [hunk 5]
```diff
@@ -47,77 +48,85 @@ final class NativePosixFileAttributeView(path: Path, options: Array[LinkOption])
...
+      private[this] var st_uid: stat.uid_t         = _
+      private[this] var st_gid: stat.gid_t         = _
+      private[this] var st_size: unistd.off_t      = _
+      private[this] var st_atime: time.time_t      = _
+      private[this] var st_mtime: time.time_t      = _
+      private[this] var st_ctime: time.time_t      = _
+      private[this] var st_blocks: stat.blkcnt_t   = _
+      private[this] var st_blksize: stat.blksize_t = _
+      private[this] var st_nlink: stat.nlink_t     = _
+      private[this] var st_mode: stat.mode_t       = _
```
Comment (densh) [inline javalib/src/main/scala/scala/scalanative/nio/fs/NativePosixFileAttributeView.scala]: Do these cached values play well with file attributes changing? i.e. what happens if file changes after the attributes where read and then the attribute info is requested later?

Comment (eatkins) [reply javalib/src/main/scala/scala/scalanative/nio/fs/NativePosixFileAttributeView.scala]: The updated semantics should be the same as on the jvm. It isn't clear from the javadoc, but when you call `Files.readAttributes`, you get a snapshot of the file attributes at the moment that the sys call to stat is invoked. For example, I ran this in the console:
```
scala> import java.nio.file._
import java.nio.file._

scala> val dir = Files.createTempDirectory("foo")
dir: java.nio.file.Path = /var/folders/7q/s3vq91lx3zv9fw4l5l5c43nr0000gn/T/foo5386655488881479548

scala> val dirAttrs = Files.readAttributes(dir, classOf[attribute.BasicFileAttributes])
dirAttrs: java.nio.file.attribute.BasicFileAttributes = sun.nio.fs.UnixFileAttributes$UnixAsBasicFileAttributes@1856eae9

scala> Files.delete(dir)

scala> Files.createFile(dir)
res1: java.nio.file.Path = /var/folders/7q/s3vq91lx3zv9fw4l5l5c43nr0000gn/T/foo5386655488881479548

scala> val attrs = Files.readAttributes(dir, classOf[attribute.BasicFileAttributes])
attrs: java.nio.file.attribute.BasicFileAttributes = sun.nio.fs.UnixFileAttributes$UnixAsBasicFileAttributes@285bb22a

scala> attrs.isRegularFile
res2: Boolean = true

scala> attrs.isDirectory
res3: Boolean = false

scala> dirAttrs.isRegularFile
res4: Boolean = false

scala> dirAttrs.isDirectory
res5: Boolean = true
```

File: javalib/src/main/scala/java/nio/file/Files.scala (hunk 1 above)
Comment (eatkins) [reply javalib/src/main/scala/java/nio/file/Files.scala]: Done.

File: javalib/src/main/scala/java/nio/file/Files.scala (hunk 2 above)
Comment (eatkins) [reply javalib/src/main/scala/java/nio/file/Files.scala]: Done.

File: javalib/src/main/scala/scala/scalanative/nio/fs/FileHelpers.scala (hunk 3 above)
Comment (eatkins) [reply javalib/src/main/scala/scala/scalanative/nio/fs/FileHelpers.scala]: Done.

File: javalib/src/main/scala/scala/scalanative/nio/fs/FileHelpers.scala (hunk 4 above)
Comment (eatkins) [reply javalib/src/main/scala/scala/scalanative/nio/fs/FileHelpers.scala]: Done.

**Comment by eatkins** [comment]
I fixed the merge conflicts and made the requested changes.

File: javalib/src/main/scala/scala/scalanative/nio/fs/NativePosixFileAttributeView.scala (hunk 5 above)
Comment (densh) [reply javalib/src/main/scala/scala/scalanative/nio/fs/NativePosixFileAttributeView.scala]: Good point, it's all good then.

**Comment by densh** [comment]
LGTM, excellent work @eatkins ! ✨🍰✨

**Pull request merged**